### ISSUE TITLE
Docs / Type Inconsistency - InMemoryCache does not conform to ApolloCacheClient / WriteData

### ISSUE THREAD

**Issue by fbartho** [apollographql/apollo-link-state#143]
The example docs for apollo-link-state tell you to pass the cache in to `withClientState(…` as part of the config object, but when you do that using an `InMemoryCache`, you get an error like the following:
```
file: 'file:///Users/${whoami}/Code/${projectName}/src/data/configureApollo.ts'
severity: 'Error'
message: 'Argument of type '{ cache: InMemoryCache; resolvers: { Mutation: any; Query: any; }; defaults: any; }' is not assignable to parameter of type 'ClientStateConfig'.
  Types of property 'cache' are incompatible.
    Type 'InMemoryCache' is not assignable to type 'ApolloCacheClient'.
      Type 'InMemoryCache' is not assignable to type 'WriteData'.
        Property 'writeData' is missing in type 'InMemoryCache'.'
at: '21,35'
source: 'ts'
code: '2345'
```
I think this is of no matter, because the cache will be passed through the context, so I *think* we shouldn't pass cache in to the configuration hash. (A reading of the unit-tests seem to back this belief up).

Either: docs & method signatures should be updated to omit `cache` from the configuration object.
Or: InMemoryCache should have the writeData method exposed so that this config is will not throw a typescript error.

```
apollo-link-state: 0.3.0
typescript: 2.6.2
```

**Comment by steve-a-jones** [comment]
@fbartho - I'm using `InMemoryCache` and everything works fine for me. Is it throwing in the latest version `0.3.0`?

In reading through https://github.com/apollographql/apollo-link-state/blob/master/packages/apollo-link-state/src/index.ts it looks like `cache` is needed when you are supplying `defaults`, otherwise it is accessed through context (as you have described).

**Comment by fbartho** [comment]
@steve-a-jones I am indeed using 0.3.0, I shall update the description to reflect that.

**Comment by Phoenixmatrix** [comment]
Just hit that issue. The problem is that `apollo-link-state` assumes a cache might not have a writeData method, and will [monkeypatch one](https://github.com/apollographql/apollo-link-state/blob/1c2d5fd275a8208b68cdab2aa32f4cdc3cc05030/packages/apollo-link-state/src/index.ts#L32) if needed.

However, the type of the config [makes writeData mandatory](https://github.com/apollographql/apollo-link-state/blob/master/packages/apollo-link-state/src/index.ts#L26) if a cache is provided at all.

So I think the types are wrong here.  `ClientStateConfig`'s cache should just be an `ApolloCache`, without the `& WriteData`, and a different type can be used in the library with `& WriteData` after writeData has been ensured/monkeypatched.

**Comment by fbartho** [comment]
Thanks for the pointer @Phoenixmatrix! I added a quick-fix for this in #157

**Comment by peggyrayzis** [comment]
Hey everyone! Thanks for reporting.

We're going to add `writeData` to the base cache type so we no longer have to monkey patch the cache, which will eliminate these errors. I'll keep you updated with when that fix is out!

**Comment by peggyrayzis** [comment]
Fixed with #164

**Comment by Phoenixmatrix** [comment]
Thanks a ton!
